cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -403,6 +444,22 @@ protected void copyArtifact(ArtifactId artifactId,
     Locations.linkOrCopy(artifactDetail.getDescriptor().getLocation(), targetFile);
   }
 
+  /**
+   * Copies the artifact jar to the given target file.
+   *
+   * @param artifactId artifact id of the artifact to be copied
+   * @param in input stream that the artifact should be copied from
+   * @param targetFile target file to copy to
+   * @throws IOException if the copying failed
+   */
+  protected void copyArtifact(ArtifactId artifactId, InputStream in, File targetFile) throws IOException {
+    try {
+      Files.copy(in, targetFile.toPath());
+    } finally {
+      in.close();
```

Comment:
usually an InputStream is closed in the same place that it is opened so that it's clear that it's closed once.

Reply:
fixed